### ISSUE TITLE
B&S assembling procedure check for position to assemble

### ISSUE THREAD

**Issue by adityaghantasala** [KratosMultiphysics/Kratos#2604]
During the assembling process in the residual_based_builder_and_solver.h specifically in the function AssembleRowContribution, there is an if condition checking if the position to assemble is less than or greater than previous position : 

https://github.com/KratosMultiphysics/Kratos/blob/59a94ce5394cee532c9ad9f0f8a33be19594ade2/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L1325

should there be check if there it is equal also ?  Something like this : 
```
            if(id_to_find > last_found)
                pos = ForwardFind(id_to_find,last_pos+1,index2_vector);
            else if(id_to_find < last_found)
                pos = BackwardFind(id_to_find,last_pos-1,index2_vector);
            else
                pos = last_pos;
```
This is necessary for the implementation of Constraints. Since when there are constraints, there is a possibility that there is a repeated entry in EquationIds.

**Comment by loumalouomega** [comment]
This is the problem earlier this morning about the need to consider all the components or it is not related?

**Comment by adityaghantasala** [comment]
@loumalouomega  yes ... this is related !!

**Comment by RiccardoRossi** [comment]
we should never have repeated entries... how can it happen?

**Comment by adityaghantasala** [comment]
for example two nodes of an element have same master node (also same DOF) .. ! . This equation number is repeated once for each slave.

**Comment by RiccardoRossi** [comment]
they should have the same EquationId in the assembled system

**Comment by adityaghantasala** [comment]
they always have ... I am saying about the elemental/conditional lhs and rhs. When two slave nodes of a condition or an element have same master node then the equation id of this particular master is repeated in EquationIds of that element/condition and the lhs and rhs have two (or more) rows and colums corresponding to that repeated master's equation id.

**Comment by RiccardoRossi** [comment]
@adityaghantasala i think i see what you are telling. to me go ahead with the change

my problem is that i don't know if this can have a performance impact on the base builder and solver

**Comment by loumalouomega** [comment]
@RiccardoRossi if the method is defined as virtual we can keep the current behavior for the base B&S and just change it in the MPC one

**Comment by RiccardoRossi** [comment]
this method @adityaghantasala is pointing to is templated, and should NOT be made virtual for performance reasons

**Comment by adityaghantasala** [comment]
resolved with #2606
